pkg/rpc/rpc_handler.go

```diff
@@ -0,0 +1,119 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package rpc
+
+import (
+	"context"
+	"net"
+
+	"github.com/aws/aws-network-policy-agent/controllers"
+	"github.com/aws/aws-network-policy-agent/pkg/utils"
+
+	"github.com/aws/amazon-vpc-cni-k8s/rpc"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/health"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/reflection"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	npgRPCaddress         = "127.0.0.1:50052"
+	grpcHealthServiceName = "grpc.health.v1.np-agent"
+)
+
+// server controls RPC service responses.
+type server struct {
+	//ebpfClient ebpf.BpfClient
+	policyReconciler *controllers.PolicyEndpointsReconciler
+	log              logr.Logger
+}
+
+// EnforceNpToPod processes CNI Enforce NP network request
+func (s *server) EnforceNpToPod(ctx context.Context, in *rpc.EnforceNpRequest) (*rpc.EnforceNpReply, error) {
+	s.log.Info("Received Enforce Network Policy Request for Pod", "Name", in.K8S_POD_NAME, "Namespace", in.K8S_POD_NAMESPACE)
+	var err error
+
+	podIdentifier := utils.GetPodIdentifier(in.K8S_POD_NAME, in.K8S_POD_NAMESPACE)
+	isMapUpdateRequired := s.policyReconciler.GeteBPFClient().IsMapUpdateRequired(podIdentifier)
+	err = s.policyReconciler.GeteBPFClient().AttacheBPFProbes(types.NamespacedName{Name: in.K8S_POD_NAME, Namespace: in.K8S_POD_NAMESPACE},
+		podIdentifier, true, true)
+	if err != nil {
+		s.log.Error(err, "Attaching eBPF probe failed for", "pod", in.K8S_POD_NAME, "namespace", in.K8S_POD_NAMESPACE)
+		return nil, err
+	}
+
+	// We attempt to program eBPF firewall map entries for this pod, if the local agent is aware of the policies
+	// configured against it. For example, if this is a new replica of an existing pod/deployment then the local
+	// node agent will have the policy information available to it. If not, we will leave the pod in default deny state
+	// until the Network Policy controller reconciles existing policies against this pod.
+
+	if s.policyReconciler.ArePoliciesAvailableInLocalCache(podIdentifier) {
```

Comment:
This can be merged - `if (isMapUpdateRequired && s.policyReconciler.ArePoliciesAvailableInLocalCache(podIdentifier))`